File: lib/vec3.lua
local class = require('lib.class')
local utils = require('lib.utils')

---@class vec3
---@overload fun(x: number?, y: number?, z: number?): vec3
---@field x number
---@field y number
---@field z number
---@operator unm:vec3
---@operator add(vec3):vec3
---@operator add(number):vec3
---@operator sub(vec3):vec3
---@operator sub(number):vec3
---@operator mul(vec3):vec3
---@operator mul(number):vec3
---@operator div(number):vec3
local vec3 = class()

---Init the vector
---@param x number?
---@param y number?
---@param z number?
function vec3:new(x, y, z)
	self.x = x or 0
	self.y = y or 0
	self.z = z or 0
end

---Set all vector values at once
---@param x number?
---@param y number?
---@param z number?
function vec3:set(x, y, z)
	self.x = x or self.x
	self.y = y or self.y
	self.z = z or self.z
end

---Replace the vector with another one
---@param new_vec3 vec3
function vec3:replace(new_vec3)
	self.x = new_vec3.x
	self.y = new_vec3.y
	self.z = new_vec3.z
end

---Clone the vector
---@return vec3
---@nodiscard
function vec3:clone()
	return vec3(self.x, self.y, self.z)
end

---Get the squared length of the vector
---@return number
---@nodiscard
function vec3:length_squared()
	return self.x * self.x + self.y * self.y + self.z * self.z
end

---Get the length of the vector
---@return number
---@nodiscard
function vec3:length()
	return math.sqrt(self:length_squared())
end

---Get the unit vector of the vector
---@return vec3
---@nodiscard
function vec3:unit_vector()
	return self / self:length()
end

---Whether the vector is very close to zero
---@return boolean
---@nodiscard
function vec3:near_zero()
	local s = 1e-8
	return (math.abs(self.x) < s) and (math.abs(self.y) < s) and (math.abs(self.z) < s)
end

---Negate the vector
---@return vec3
---@nodiscard
function vec3:__unm()
	return vec3(-self.x, -self.y, -self.z)
end

---Add two vectors or a vector and a number
---@param a vec3|number
---@param b vec3|number
---@return vec3
---@nodiscard
function vec3.__add(a, b)
	if type(a) == 'number' then
		-- if `a` is a number then `b` has to be vec3
		return vec3(a + b.x, a + b.y, a + b.z)
	elseif type(b) == 'number' then
		-- if `b` is a number then `a` has to be vec3
		return vec3(a.x + b, a.y + b, a.z + b)
	end

	return vec3(a.x + b.x, a.y + b.y, a.z + b.z)
end

---Subtract two vectors or a vector and a number
---@param a vec3|number
---@param b vec3|number
---@return vec3
---@nodiscard
function vec3.__sub(a, b)
	if type(a) == 'number' then
		-- if `a` is a number then `b` has to be vec3
		return vec3(a - b.x, a - b.y, a - b.z)
	elseif type(b) == 'number' then
		-- if `b` is a number then `a` has to be vec3
		return vec3(a.x - b, a.y - b, a.z - b)
	end

	return vec3(a.x - b.x, a.y - b.y, a.z - b.z)
end

---Multiply two vectors or a vector and a number
---@param a vec3|number
---@param b vec3|number
---@return vec3
---@nodiscard
function vec3.__mul(a, b)
	if type(a) == 'number' then
		-- if `a` is a number then `b` has to be vec3
		return vec3(a * b.x, a * b.y, a * b.z)
	elseif type(b) == 'number' then
		-- if `b` is a number then `a` has to be vec3
		return vec3(a.x * b, a.y * b, a.z * b)
	end

	return vec3(a.x * b.x, a.y * b.y, a.z * b.z)
end

---Divide a vector and a number
---@param a vec3|number
---@param b vec3|number
---@return vec3
---@nodiscard
function vec3.__div(a, b)
	if type(a) == 'number' then
		-- if `a` is a number then `b` has to be vec3
		return (1 / a) * b
	elseif type(b) == 'number' then
		-- if `b` is a number then `a` has to be vec3
		return a * (1 / b)
	end

	error('Invalid vec3 division: ' .. type(a) .. ' / ' .. type(b))
end

---Return the dot product of two vectors
---@param a vec3
---@param b vec3
---@return number
---@nodiscard
function vec3.dot(a, b)
	return a.x * b.x + a.y * b.y + a.z * b.z
end

---Return the cross product of two vectors
---@param a vec3
---@param b vec3
---@return vec3
---@nodiscard
function vec3.cross(a, b)
	return vec3(a.y * b.z - a.z * b.y, a.z * b.x - a.x * b.z, a.x * b.y - a.y * b.x)
end

---Reflect a vector
---@param v vec3
---@param n vec3
---@return vec3
---@nodiscard
function vec3.reflect(v, n)
	return v - n * v:dot(n) * 2
end

---Refract a vector
---@param uv vec3
---@param n vec3
---@param etai_over_etat number
---@return vec3
---@nodiscard
function vec3.refract(uv, n, etai_over_etat)
	local cos_theta = math.min((-uv):dot(n), 1.0)
	local r_out_perp = (uv + n * cos_theta) * etai_over_etat
	local r_out_parallel = n * -math.sqrt(math.abs(1.0 - r_out_perp:length_squared()))
	return r_out_perp + r_out_parallel
end

---Get a random vector
---@param min number?
---@param max number?
---@return vec3
---@nodiscard
function vec3.random(min, max)
	if min and max then
		return vec3(utils.random_float(min, max), utils.random_float(min, max), utils.random_float(min, max))
	end

	return vec3(math.random(), math.random(), math.random())
end

---Get a random vector in unit sphere
---@return vec3
---@nodiscard
function vec3.random_in_unit_sphere()
	while true do
		local x = utils.random_float(-1, 1)
		local y = utils.random_float(-1, 1)
		local z = utils.random_float(-1, 1)

		if x * x + y * y + z * z < 1 then
			return vec3(x, y, z)
		end
	end
end

---Get a random unit vector
---@return vec3
---@nodiscard
function vec3.random_unit_vector()
	return vec3.random_in_unit_sphere():unit_vector()
end

---Get a random vector in unit disk
---@return vec3
---@nodiscard
function vec3.random_in_unit_disk()
	while true do
		local x = utils.random_float(-1, 1)
		local y = utils.random_float(-1, 1)

		if x * x + y * y < 1 then
			return vec3(x, y, 0)
		end
	end
end

return vec3

File: src/vec3.ts
import type { FileHandle } from "node:fs/promises";

export class Vec3 {

	public constructor(
		public x = 0,
		public y = 0,
		public z = 0,
	) {}

	public add(v: Vec3) {
		this.x += v.x;
		this.y += v.y;
		this.z += v.z;
		return this;
	}

	public static add(a: Vec3, b: Vec3) {
		return new Vec3(
			a.x + b.x,
			a.y + b.y,
			a.z + b.z,
		);
	}

	public sub(v: Vec3) {
		this.x -= v.x;
		this.y -= v.y;
		this.z -= v.z;
		return this;
	}

	public static sub(a: Vec3, b: Vec3) {
		return new Vec3(
			a.x - b.x,
			a.y - b.y,
			a.z - b.z,
		);
	}

	public mul(v: Vec3) {
		this.x *= v.x;
		this.y *= v.y;
		this.z *= v.z;
		return this;
	}

	public static mul(a: Vec3, b: Vec3) {
		return new Vec3(
			a.x * b.x,
			a.y * b.y,
			a.z * b.z,
		);
	}

	public scale(t: number) {
		this.x *= t;
		this.y *= t;
		this.z *= t;
		return this;
	}

	public static scale(v: Vec3, t: number) {
		return new Vec3(
			v.x * t,
			v.y * t,
			v.z * t,
		);
	}

	public down(t: number) {
		return this.scale(1 / t);
	}

	public static down(v: Vec3, t: number) {
		return Vec3.scale(v, 1 / t);
	}

	public dot(b: Vec3) {
		return this.x * b.x + this.y * b.y + this.z * b.z;
	}

	public static dot(a: Vec3, b: Vec3) {
		return a.x * b.x + a.y * b.y + a.z * b.z;
	}

	public cross(v: Vec3) {
		const x = this.x;
		const y = this.y;
		const z = this.z;
		const vx = v.x;
		const vy = v.y;
		const vz = v.z;
		this.x = y * vz - z * vy;
		this.y = z * vx - x * vz;
		this.z = x * vy - y * vx;
		return this;
	}

	public static unitVector(v: Vec3) {
		return Vec3.down(v, v.length);
	}

	public get length() {
		return Math.sqrt(this.lengthSquared);
	}

	public get lengthSquared() {
		return this.x * this.x + this.y * this.y + this.z * this.z;
	}

	public toString() {
		return `{${Math.floor(this.x*1000)/1000}, ${Math.floor(this.y*1000)/1000}, ${Math.floor(this.z*1000)/1000}}`;
	}

	public async writeAsColorToFile(fh: FileHandle) {
		await fh.write(`${Math.trunc(255.999 * this.x)} ${Math.trunc(255.999 * this.y)} ${Math.trunc(255.999 * this.z)}\n`);
	}

}

export { Vec3 as Point3 };
export { Vec3 as Color };
